Protobuf updates

kgururaj opened the pull request:
This PR implements populating a GenomicsDBConfigBase object from a Protobuf ExportConfiguration object. This can be used when transferring information across languages (Python/R/Java -> C++). You can serialize the Protobuf object into a string buffer (JSON or binary format) from the high-level language side, pass the serialized byte buffer to C++ where it's deserialized.

[Example/test code](https://github.com/kgururaj/GenomicsDB/blob/protobuf_updates/src/test/cpp/src/test_pb.cc#L58)

nalinigans commented on src/test/cpp/src/ctest_main.cc:
```diff
@@ -24,6 +24,39 @@
+{
+  Catch::Session session; // There must be exactly one instance
+
+  int height = 0; // Some user variable you want to be able to set
```
Looks like cruft from the Catch2 docs. Is it needed here?

kgururaj replied:
Removed

nalinigans commented:
@kgururaj, the travis build for this PR is failing for MacOS - the protobuf version is 3.0.0-beta-1. 

```
/Users/travis/build/GenomicsDB/GenomicsDB/src/test/cpp/src/test_pb.cc:71:43: error: no member named 'JsonStringToMessage' in namespace 'google::protobuf::util'
    auto status = google::protobuf::util::JsonStringToMessage(json_buffer, &export_config);
                  ~~~~~~~~~~~~~~~~~~~~~~~~^
1 error generated.
```
What are your recommendations for the protobuf version that also is compatible with gatk/hdfs?

mlathara commented on src/main/cpp/include/config/pb_config.h:
```diff
@@ -0,0 +1,41 @@
+#include "genomicsdb_config_base.h"
+#include "genomicsdb_export_config.pb.h"
+
+class PBConfigBase : public GenomicsDBConfigBase {
```
This is probably a dumb question, but why is this GenomicsDBConfigBase here instead of VariantQueryConfig?

kgururaj replied:
Not a dumb question  by any means - I have struggled to determine what's the best class hierarchy here (or whether it's even needed)
```suggestion
A<--B  B is a subclass of A
GenomicsDBConfigBase <-- JSONConfigBase
GenomicsDBConfigBase <-- PBConfigBase
GenomicsDBConfigBase <-- GenomicsDBImportConfig
GenomicsDBConfigBase <-- VariantQueryConfig
```
So, you could create a PBConfigBase object and then:
```
  PBConfigBase pb_cfg;
  //initializes struct here and then ..
  VariantQueryConfig qcfg;
  static_cast<GenomicsDBConfigBase&>(qcfh) = std::move(static_cast<GenomicsDBConfigBase&>(pb_cfg));
```
An alternative is to get rid of JSONConfigBase and PBConfigBase and move the functions to GenomicsDBConfigBase (read_from_JSON_file, read_from_PB). Would probably simplify matters.

nalinigans replied:
> static_cast<GenomicsDBConfigBase&>(qcfh) = std::move(static_cast<GenomicsDBConfigBase&>(pb_cfg));
>An alternative is to get rid of JSONConfigBase and PBConfigBase and move the functions to GenomicsDBConfigBase (read_from_JSON_file, read_from_PB). 
Yes, this would simplify things as reading from JSON/protobuf files would be part of the base configuration, e.g. specifying workspace would be treated the same whether it came from GenomicsDBImportConfig api directly or JSON files or protobuf. It would also avoid moving JSON|PBConfigBase properties back into VariantQueryConfig in a roundabout way.

nalinigans replied:
One other question not related to this PR and to just increase my understanding, why is the bookkeeping operation not part of instantiating VariantQueryProcessor? Should it remain separate or can we merge it in?

kgururaj replied:
The bookkeeping depends on the input query (JSON/Protobuf) of the user (fields, bounds etc).

mlathara commented on src/main/cpp/src/config/pb_config.cc:
```diff
@@ -0,0 +1,230 @@
+      m_column_ranges.resize(export_config->query_column_ranges_size());
+      for(auto i=0;i<export_config->query_column_ranges_size();++i) {
+	const auto& inner_list_wrapper = export_config->query_column_ranges(i);
+	m_column_ranges[i].resize(inner_list_wrapper.column_or_interval_list_size());
```
Maybe inline with the earlier question about GenomicsDBConfigBase vs VariantQueryConfig...if we did use VariantQueryConfig as the parent we could directly initialize m_query_column_intervals here instead of m_column_ranges, right? And similarly for m_query_rows and m_row_ranges below....

kgururaj replied:
I use the GenomicsDBConfigBase class as a 'bag' for the information provided by JSON/Protobuf (both import and query). The VariantQueryConfig members are somewhat curated - example, if the array is partitioned, then the m_query_column_intervals are cut at the partition bounds.
But, yes, you could rewrite the structures to avoid the duplication.

kgururaj commented:
> @kgururaj, the travis build for this PR is failing for MacOS - the protobuf version is 3.0.0-beta-1.
> 
> ```
> /Users/travis/build/GenomicsDB/GenomicsDB/src/test/cpp/src/test_pb.cc:71:43: error: no member named 'JsonStringToMessage' in namespace 'google::protobuf::util'
>     auto status = google::protobuf::util::JsonStringToMessage(json_buffer, &export_config);
>                   ~~~~~~~~~~~~~~~~~~~~~~~~^
> 1 error generated.
> ```
> 
> What are your recommendations for the protobuf version that also is compatible with gatk/hdfs?

Fixed - I added a comment to explain why this failed. A roundabout way is used in the test to work with 3.0.0-beta-1.

kgururaj commented on src/resources/genomicsdb_export_config.proto:
```diff
@@ -47,20 +49,21 @@ message RowRangeList {
     bool generate_array_name_from_partition_bounds = 18;
   }
-  required string reference_genome = 3;
+  optional string reference_genome = 3;
```
The reference_genome is only required if you are generating VCFs

nalinigans replied:
Ah, thanks. I will have to reflect this in the api, where it is required currently.
